### ISSUE TITLE
Potential bug about book_order

### ISSUE THREAD

**Issue by zeyuhuan** [nautechsystems/nautilus_trader#1864]
# Bug Report
I have a use case to create a list of OrderBook from snapshot data and feed to DataEngine. And when I create order book, I encountered potential memory leak. Here is the code that should be reproducible with memory leak

```
from nautilus_trader.model.enums import OrderSide, BookType
from nautilus_trader.model.book import OrderBook
from nautilus_trader.model.objects import Price, Quantity
from nautilus_trader.model.data import DataType, CustomData, DataType, BookOrder
from nautilus_trader.model.identifiers import InstrumentId

from trade.strategies.utils import symbol_to_instrument_id


def symbol_to_binance_symbol(symbol: str) -> str:
    return symbol.replace("-", "")


def symbol_to_instrument_id(symbol: str) -> InstrumentId:
    return InstrumentId.from_str(f"{symbol_to_binance_symbol(symbol)}.BINANCE")


def df_to_books(_: None, instrument_id_map):
    books = []
    data_type = DataType(OrderBook)

    for _ in range(100000):
        ask_prices = range(10)
        ask_sizes = range(10)
        bid_prices = range(10)
        bid_sizes = range(10)
        symbol = "BTC-USDT"
        order_book = OrderBook(instrument_id_map[symbol], BookType.L2_MBP)
        timestamp = 11111111111
        sequence = 1
        for price, size in zip(ask_prices, ask_sizes):
            book_order = BookOrder(
                OrderSide.SELL,
                Price(price, 9),
                Quantity(size, 9),
                sequence,
            )
            order_book.add(book_order, timestamp)
        for price, size in zip(bid_prices, bid_sizes):
            book_order = BookOrder(
                OrderSide.BUY,
                Price(price, 9),
                Quantity(size, 9),
                sequence,
            )
            order_book.add(book_order, timestamp)

        sequence += 1
        books.append(CustomData(data_type, order_book))

    return books


instrument_id_map = {
    symbol: symbol_to_instrument_id(symbol) for symbol in ("BTC-USDT", "ETH-USDT")
}

for _ in range(100000):
    processed_book = df_to_books(_, instrument_id_map)
    del processed_book
```
when I run this code, memory keeps increasing

and when I comment out this the two lines of `order_book.add(book_order, timestamp)`, memory doesn't increase when I run the code

### Expected Behavior
Memory usage should not increase, as memory used from  `df_to_books` should be reclaimed in each for loop

### Actual Behavior
Memory usage keeps increasing

### Steps to Reproduce the Problem

1. install nautilus_trader and run the code snippet I provided above, monitor memory and it keeps increasing
2. comment out two lines of `order_book.add(book_order, timestamp)`, monitor memory, it does't increase

### Specifications

- OS platform: Ubuntu 22
- Python version: 3.10.14
- `nautilus_trader` version: 1.192.0 / 1.199.0

**Comment by zeyuhuan** [comment]
Looking at `book.rs`, looks like I need `orderbook_drop` expose to python so that I can free the mem. But I am not sure if this the best fix

**Comment by cjdsellers** [comment]
Hey @zeyuhuan

So the way you're accessing the order book here is quite low level. You're bypassing the order book deltas as you mentioned and just adding more and more orders to the book (this is why memory increases, as the book will hang onto those references until the orders are removed). The order book deltas have an `action` field which specifies how they should update the book state, including updates and deletes.

If you're using Binance data then I'd recommend following the tutorial which keeps usage up at the intended API level
https://nautilustrader.io/docs/latest/tutorials/backtest_binance_orderbook.

I hope that helps!

Closing this for now, but feel free to follow up with any thoughts or updates here.

**Comment by cjdsellers** [comment]
Hi @zeyuhuan

Looking more closely, I see you're referring to the book itself and not just the book orders.

It turns our you're correct, there was actually a bug where the backing struct was not being freed with `orderbook_drop` as part of the object destruction in the Python finalizer, causing a memory leak. Since Python is a managed language, we generally avoid exposing memory allocation functions to the Python API. Instead, we should leverage Python's standard data model and rely on its garbage collector to handle memory management.

Now fixed from a0694c616b8b164d0490931614616b6d245593f4 and credited your report in the release notes!
